src/coreclr/jit/optimizer.cpp

```diff
@@ -6382,17 +6366,26 @@ bool Compiler::optHoistThisLoop(FlowGraphNaturalLoop* loop, LoopHoistContext* ho
                 ", or pre-headers of nested loops, if any:\n",
                 exiting->bbNum);
 
-        // Push dominators, until we reach "entry" or exit the loop.
-
+        // Push dominators, until we reach the header or exit the loop.
+        //
+        // Note that there is a mismatch between the dominator tree dominance
+        // and loop header dominance; the dominator tree dominance relation
+        // guarantees that a block A that dominates B was exited before B is
+        // entered, meaning it could not possibly have thrown an exception. On
+        // the other hand loop finding guarantees only that the header was
+        // entered before other blocks in the loop. If the header is a
+        // try-begin then blocks inside the catch may not necessarily be fully
+        // dominated by the header, but may still be part of the loop.
```

Comment:
This is something we could consider canonicalizing, though I'm not sure it is really necessary (I would expect most reasoning about the header to already take into account that only the "beginning" of it is guaranteed to be executed).

An example that hits the assert in the base looks like:
```csharp
private static int Foo(int[] arr, int n)
{
    int sum = 0;
    for (int i = 0; i < 100; i++)
    {
        try
        {
            sum += arr[i];
        }
        catch (IndexOutOfRangeException)
        {
        }
    }

    return sum;
}
```
I think we should be able to recognize and optimize loops like this. The flow graph looks like [this](https://edotor.net/?engine=dot?engine=dot#digraph%20FlowGraph%20%7B%0A%20%20%20%20graph%20%5Blabel%20%3D%20%22Foo%5Cnafter%5CnDo%20value%20numbering%22%5D%3B%0A%20%20%20%20node%20%5Bshape%20%3D%20%22Box%22%5D%3B%0A%20%20%20%20BB01%20%5Blabel%20%3D%20%22BB01%5CnV03%20%3E%3D%20100%22%2C%20shape%20%3D%20%22house%22%5D%3B%0A%20%20%20%20BB02%20%5Blabel%20%3D%20%22BB02%22%2C%20shape%20%3D%20%22note%22%5D%3B%0A%20%20%20%20BB03%20%5Blabel%20%3D%20%22BB03%22%5D%3B%0A%20%20%20%20BB04%20%5Blabel%20%3D%20%22BB04%5CnV03%20%3C%20100%22%5D%3B%0A%20%20%20%20BB05%20%5Blabel%20%3D%20%22BB05%22%2C%20shape%20%3D%20%22invhouse%22%5D%3B%0A%20%20%20%20BB06%20%5Blabel%20%3D%20%22BB06%22%5D%3B%0A%20%20%20%20BB01%20-%3E%20BB02%20%5Bcolor%3Dblue%2C%20weight%3D20%5D%3B%0A%20%20%20%20BB02%20-%3E%20BB03%20%5Bcolor%3Dblue%2C%20weight%3D20%5D%3B%0A%20%20%20%20BB04%20-%3E%20BB03%20%5Bcolor%3Dgreen%5D%3B%0A%20%20%20%20BB03%20-%3E%20BB04%20%5Bcolor%3Dblue%2C%20weight%3D20%5D%3B%0A%20%20%20%20BB06%20-%3E%20BB04%20%5Bcolor%3Dgreen%5D%3B%0A%20%20%20%20BB01%20-%3E%20BB05%20%5B%5D%3B%0A%20%20%20%20BB04%20-%3E%20BB05%20%5Bcolor%3Dblue%2C%20weight%3D20%5D%3B%0A%20%20%20%20BB01%20-%3E%20BB02%20%5Bstyle%3D%22invis%22%2C%20weight%3D25%5D%3B%0A%20%20%20%20BB02%20-%3E%20BB03%20%5Bstyle%3D%22invis%22%2C%20weight%3D25%5D%3B%0A%20%20%20%20BB03%20-%3E%20BB04%20%5Bstyle%3D%22invis%22%2C%20weight%3D25%5D%3B%0A%20%20%20%20BB04%20-%3E%20BB05%20%5Bstyle%3D%22invis%22%2C%20weight%3D25%5D%3B%0A%20%20%20%20BB05%20-%3E%20BB06%20%5Bstyle%3D%22invis%22%2C%20weight%3D25%5D%3B%0A%20%20%20%20subgraph%20cluster_0%20%7B%0A%20%20%20%20%20%20%20%20label%20%3D%20%22L00%22%3B%0A%20%20%20%20%20%20%20%20color%20%3D%20blue%3B%0A%20%20%20%20%20%20%20%20BB03%3BBB06%3BBB04%3B%0A%20%20%20%20%7D%0A%7D%0A):
```scala
-----------------------------------------------------------------------------------------------------------------------------------------
BBnum BBid ref try hnd preds           weight    lp [IL range]     [jump]      [EH region]         [flags]
-----------------------------------------------------------------------------------------------------------------------------------------
BB01 [0000]  1                             1       [000..006)-> BB05 ( cond )                     i 
BB02 [0008]  1       BB01                  1       [006..???)-> BB03 (always)                     internal LoopPH q 
BB03 [0002]  2  0    BB02,BB04             4       [006..00F)-> BB04 (always) T0      try { }     i keep Loop idxlen bwd q 
BB04 [0004]  2       BB03,BB06             8       [012..01B)-> BB03 ( cond )                     i bwd 
BB05 [0006]  2       BB01,BB04             1       [01B..01D)        (return)                     i 
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++ funclets follow
BB06 [0003]  1     0                       0       [00F..012)-> BB04 ( cret )    H0 F catch { }   i rare keep xentry flet bwd 
-----------------------------------------------------------------------------------------------------------------------------------------

...

L00 header: BB03
  Members (3): [BB03..BB04];BB06
  Entry: BB02 -> BB03
  Exit: BB04 -> BB05
  Back: BB04 -> BB03

```

`BB06` is the catch block; it is part of the loop but its immediate dominator is `BB02`, since `BB03` (the header) is not guaranteed to be exited before `BB06` is entered.

I think loops like these are definitely ones we want to be able to recognize and handle. If we run into more odd special casing then I think we can canonicalize these cases by introducing a block before the "try" begin, so that the try begin does not become the header. (Note that the loop-inside-try could also have the try-begin as the header, but would not have the catch blocks considered as part of the loop)

Reply:
> I think we can canonicalize these cases by introducing a block before the "try" begin, so that the try begin does not become the header

I like that idea. Removing cases where we pessimize due to difficult EH flow graph structures is a good thing.